testsuite/forge/src/success_criteria.rs

```diff
@@ -76,6 +89,98 @@ impl SuccessCriteria {
                 .await?;
         }
 
+        if let Some(chain_progress_threshold) = &self.chain_progress_check {
+            let epochs = FetchMetadata::fetch_new_block_events(
+                swarm.aptos_public_info().client(),
+                None,
+                None,
+            )
+            .await
+            .unwrap();
+
+            let mut max_round_gap = 0;
+            let mut max_round_gap_version = 0;
+            let mut max_time_gap = 0;
+            let mut max_time_gap_version = 0;
+
+            let mut prev_block = None;
+            let mut prev_ts = 0;
+            let mut failed_from_nil = 0;
+            let mut previous_epooch = 0;
+            let mut previous_round = 0;
+            for block in epochs
+                .iter()
+                .flat_map(|epoch| epoch.blocks.iter())
+                .filter(|b| b.version > start_version && b.version < end_version)
+            {
+                let is_nil = block.event.proposer() == PeerId::ZERO;
+
+                let current_gap = if previous_epooch == block.event.epoch() {
+                    block.event.round() - previous_round - 1
+                } else {
+                    (if is_nil { 0 } else { 1 })
+                        + block.event.failed_proposer_indices().len() as u64
+                };
+
+                if is_nil {
```

Comment:
On a high level, this check only checks if the network is progressing by producing TC, it is possible that no TC is produced and so all nodes continuously timeout. Any thoughts on how we can check that?

Reply:
I am ignoring nil blocks,  basically treating that as a gap.

This doesn't check if network is still halted,  maybe a timestamp check would be enough here

Reply:
Is your `max_no_progress_secs` going to catch that?

Reply:
No,  because I check gaps between blocks.  If NewBlockEvents just permanently stop, there is no "last" one to compare to